### DISCUSSION TITLE
Firebase CLI configuee multiple flavor

### DISCUSSION THREAD

**Discussion by HoangNguyen0403** [firebase/flutterfire#8691] [General]
Hello guys, i found that we can customize package name and output file to generate multiple flavor for firebase config. But with dsyms on IOS crashlytics it'll be sended to firebase with firebase_app_id_file.json that will be generated one time per flavors. How can we customize this file output to multiple flavor ?

**Comment by ghost** [comment]
I would also love to hear how others are accomplishing this.  Currently my plan is to create either a prebuild script to copy the env version to override the `firebase_app_id_file.json` version.  I think there is also a way to copy it in build phases to the correct location as well but I have less experience with that.  I wish flutterfire cli would handle this, flavors have been a part of flutter for some time now.

**Reply by edwardmp** [reply]
You can do this, but please note that this json file is never used, except if you configure Crashlytics. So otherwise don't bother with it.

Source: https://github.com/firebase/flutterfire/blob/master/packages/firebase_crashlytics/firebase_crashlytics/ios/crashlytics_add_upload_symbols#L44

**Comment by rafoolin** [comment]
Hi, similar to `GoogleService-Info.plist` files for each flavor, we did the same for `firebase_app_id_file.json`. You can follow [this tutorial](https://medium.com/@matt.goodson.business/separating-build-environment-configurations-in-flutter-with-firebase-doing-it-the-right-way-c72c3ad3621f) and check the script that does this.

**Reply by HoangNguyen0403** [reply]
Thank you! Let me try with this way and get back result

**Reply by HoangNguyen0403** [reply]
After checking this tutorial. It's similar to the old manual configuration firebase multiple flavors. What I want here is using firebase_cli to generate multiple google-service.json files for android and GoogleService-Info.plist on other destination folders. For now, it's only one folder when generating google-service.json with firebase configuration. I want to have another option like "-out" for firebase_option.dart file.

**Reply by HoangNguyen0403** [reply]
For example: "flutterfire configure -i "your-bundle-id" -a "your-package-name" -o
lib/firebase/firebase_options_dev.dart" - android_service "android/app/dev/google-services.json" -iso_service "ios/dev/firebase_app_id_file.json"

**Reply by SharbelOkzan** [reply]
This tutorial saved me, thanks! This was the best approach to follow.

Generate the files using `flutterfire` CLI for each flavor independently, move the generated `.json` and `.plist` files to a directory you define, lastly, follow the above tutorial to create a prebuild xcode script to copy the files back to the appropriate destination. (it's simpler for Android, and also included in the tutorial)

**Comment by HoangNguyen0403** [comment]
Here I bring my idea and suggestion about adding more config for google_service.json and google_service.plist output for support of multiple flavors.

Usage: flutterfire configure [arguments]
-h, --help                                  Print this usage information.
-p, --project=<aliasOrProjectId>            The Firebase project to use for this command.
-e, --account=<email>                       The Google account to use for authorization.
-o, --out=<filePath>                        The output file path of the Dart file that will be generated with your Firebase configuration options.
                                            (defaults to "lib/firebase_options.dart")
**-ao, —android-out=<filePath>                        The output file path of the google_service.json file that will be generated with your Firebase configuration options.
                                            (defaults to "android/app”) Ex: “android/app/dev”
-io, —ios-out=<filePath>                        The output file path of the Google_Service.plist file that will be generated with your Firebase configuration options.
                                            (defaults to “ios/”) Ex: “iOS/dev”**
-y, --yes                                   Skip the Y/n confirmation prompts and accept default options (such as detected platforms).
    --platforms=<platforms>                 Optionally specify the platforms to generate configuration options for as a comma separated list. For example
                                            "android,ios,macos,web,linux,windows".
-i, --ios-bundle-id=<bundleIdentifier>      The bundle identifier of your iOS app, e.g. "com.example.app". If no identifier is provided then an attempt will be made to automatically
                                            detect it from your "ios" folder (if it exists).
-m, --macos-bundle-id=<bundleIdentifier>    The bundle identifier of your macOS app, e.g. "com.example.app". If no identifier is provided then an attempt will be made to
                                            automatically detect it from your "macos" folder (if it exists).
    --android-app-id=<applicationId>        DEPRECATED - use "android-package-name" instead. The application id of your Android app, e.g. "com.example.app". If no identifier is
                                            provided then an attempt will be made to automatically detect it from your "android" folder (if it exists)
-a, --android-package-name=<packageName>    The package name of your Android app, e.g. "com.example.app". If no package name is provided then an attempt will be made to
                                            automatically detect it from your "android" folder (if it exists).

**Comment by iamnabink** [comment]
Just following it up ✌🏼

**Comment by OutdatedGuy** [comment]
Same problem with me. I want to use 2 different firebase projects simultaneously in my flutter app.
I'm using the below command to add project configs into my flutter app:

```bash
flutterfire configure --project="proj1" --platforms="android,ios" --out="lib/firebase_options.dart" -y;
flutterfire configure --project="proj2" --platforms="android,ios" --out="lib/proj2.firebase_options.dart";
```

and want to configure them at the same time with the below code in the main function:

```dart
import 'firebase_options.dart' as proj1;
import 'proj2.firebase_options.dart' as proj2;

void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await Future.wait([
    Firebase.initializeApp(
      options: proj1.DefaultFirebaseOptions.currentPlatform,
    ),
    Firebase.initializeApp(
      name: "Project2",
      options: proj2.DefaultFirebaseOptions.currentPlatform,
    ),
  ]);

  runApp(MyApp());
}

``` 

<hr />

But when I'm trying to run the configure commands it is asking if I want to override the `firebase_app_id_file.dart` as it is already configured for project1.

When I type `no` (cuz proj1 is main one), I get the following error:

```bash
Unhandled exception:
FirebaseAppIDAlreadyExistsException: Firebase app ID file {{project_path}}/ios/firebase_app_id_file.json is already up to date.
#0      FirebaseAppIDFile.write (package:flutterfire_cli/src/firebase/firebase_app_id_file.dart:71:11)
<asynchronous suspension>
```

**Reply by gem-harshitsrivastava** [reply]
Just wanted to ask that even though we are can use the two firebase apps for same applications different flavors but what about the SHA keys which can only be applied to either of them and sha keys are mandatory for auth services.

**Reply by OutdatedGuy** [reply]
SHA 1 is only needed for Auth or dynamic links, and I'm not planning to use 2nd project for auth but just database. So I'm fine in that case.

**Reply by OutdatedGuy** [reply]
Ohhh, I think I misread your question. I don't know about flavour and auth using 2 firebase projects with same app.

**Reply by edwardmp** [reply]
@gem-harshitsrivastava you can use the same SHA keys multiple times, but you need to make sure that your flavors have different bundle ids. E.g. `com.mycompany.myapp` and `com.mycompany.myapp.sandbox` is what I use. It works fine.

**Comment by ycv005** [comment]
Just following it up ✌🏼

**Comment by amialif** [comment]
so still this cli can't configure the .json and .plist file ?  need to go back to that manual setup  ?

**Comment by wildsurfer** [comment]
+1 for clear instructions on how to use flutterfire with flavours.

**Comment by edlman** [comment]
I was using flavors in Flutter without problem with gradle plugin 4.3.x, where each flavor has its own `google-services.json` downloaded from Firebase console. Plugin version 4.4.x seems not to use `google-services.json` files but require Dart file config generated with `flutterfire configure` command. This is not very handy for flavors.
So I downgraded gradle plugin back to 4.3 version in `android/build.gradle`, `buildscript { dependencies { classpath 'com.google.gms:google-services:4.3.15' }}` and flavor based configs in `google-services.json` work.

**Comment by ghost** [comment]
Following up on this also. Seems like it should be reasonable to do, but no accepted method.

**Comment by wiesnery** [comment]
Would love to see a response here as well.
It is confusing, that plugins like flutter_flavorizr are keeping helping out with this but no official guidance is given.

Could you in the meanwhile help me out with one question:
flutter_flavorizr is helping to flavor the google-services.json and GoogleService-Info.plist but does not include the firebase_app_id_file.json
I am using mainly Crashlytics - is the id_file still necessary? Is it better to use the google-services.json and GoogleService-Info.plist which flutter_flavorizr handles nicely or is configuration outdated and should be done via the dart files?

**Reply by bziembanski** [reply]
Hi, any follow up on that? Would love to see any possibility to still use google-services.json files due to using flavors

**Reply by wiesnery** [reply]
I think a quick answer on this question would be helpful:

Is the firebase_app_id_file.json necessary when platform specific google-services.json and GoogleService-Info.plist are used with Flutter?

Any engineer of the framework here, who could help with this?
